[grpc/protobuf] Update grpc to 1.60.0 and update protobuf to 3.25.1

Tradias opened the pull request:
Resolves https://github.com/microsoft/vcpkg/issues/35566

This pr includes https://github.com/microsoft/vcpkg/pull/31159 and https://github.com/microsoft/vcpkg/pull/35399

Ports changed in this pr:

Most of these changes stem from the fact that protobuf now depends on abseil and requires c++14 while ports consume protobuf using `target_link_libraries(lib ${Protobuf_LIBRARIES})` instead of `target_link_libraries(lib PUBLIC protobuf::libprotobuf)`.

* **abseil** Updated to 03/04/2024 to address MSVC build issue in openvino
* **arcus** Updated to 4.13.2
* **braft** Patched to use `find_package(Protobuf` instead of custom FindProtobuf module. also link with `PUBLIC protobuf::libprotobuf)`.
* **brpc**  Patch several `target_link_library` calls to include `PUBLIC` instead of nothing. Patch some warnings that are treated as errors by some OSX compiler. Patch usage of changed protobuf features most importantly removal of `SetLogHandler`, tbd whether this patch is acceptable with upstream.
* **cld3** Use CONFIG to find protobuf to propagate dependent abseil libs and cxx14 correctly.
* **ecal** Use CONFIG to find protobuf. Use PUBLIC when linking it.
* **gamenetworkingsockets** Add `-DCMAKE_FIND_PACKAGE_PREFER_CONFIG=on` for protobuf.
* **gz-transport12** Update to fix usage of removed protobuf features.
* **ignition-msgs1/5/6** Remove protobuf version check which is confused by the duplicate versioning scheme (v25.1 vs. 3.25.1 vs 4.x). Why are we even maintaining this no longer maintained major version of this port? I would welcome a more strict policy on removing ports/adding them to ci.baseline.txt
* **ignition-transport4/8/9** Same as above + cherry-pick a protobuf deprecation patch.
* **libprotobuf-mutator** Cherry-pick two commits to address altered protobuf functionality. Add CONFIG and PUBLIC to protobuf handling in CMake.
* **marble** Explicitly set protobuf to not found since it is not part of the vcpkg.json but is made available transitively but another dependency.
* **mysql-connector-cpp** Several CMake changes and removal of protobuf::SetLogHandler patch.
* **openvino** Find protobuf using CONFIG
* **osgearth** Link libraries using PUBLIC and link with protobuf::libprotobuf instead of Protobuf_LIBRARIES
* **paraview** Find protobuf using CONFIG and remove version check
* **pulsar-client-cpp** Add protobuf linkage to PULSAR_OBJECT_LIB. Simplify protoc patch. Remove -Werror.
* **shogun** Update and patch as much as possible. Users are required to override bitsery version to 4.x to use shogun since it does not support 5.x provided by vcpkg. Therefore add shogun to ci.baseline.txt.
* **srpc** Update and fix static crt linkage. Patch protobuf linkage and remove hardcoded `-std=c++11` flag. Protobuf requires c++14 and CMake does not recognize the hardcoded c++11 flag and will therefore not add a c++14 flag when the compiler uses c++14 by default.
*  **upb** Update to align version with protobuf/grpc and use new github repo. Patch usual CMake mess (they do not support CMake officially) which should probably be `unofficial-upbConfig.cmake` but I didn't dig into downstream implications so I kept it at `upbConfig.cmake` as before.
* **utf8-range** Update to allign version with protobuf.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

Tradias committed:
grpc: Update to 1.60.0. Update protobuf, upb and utf8-range to 3.25.1

Tradias committed:
protobuf: Fix dll macro and add patches for libprotobuf-mutator. Also ignore mysql-connector-cpp in ci.baseline

Tradias committed:
protobuf: Fix tool installation for non-windows

Tradias committed:
grpc: Fix pluginTargets.cmake. Also update gz-transport12 and patch openvino to establish protobuf v25.1 compat

Tradias committed:
protobuf: Fix several dependent packages after major version update and C++14 requirement

Tradias committed:
protobuf: Fix dependent ignition-transport, gamenetworkingsockets and libosmscout ports

Adela0814 commented:
Pinging @Tradias  for response. Is work still being done for this PR?

Tradias committed:
protobuf: Update and patch brpc and shogun. Simplify paraview protobuf patch

Tradias committed:
protobuf: Update arcus. Fix libprotobuf-mutator protobuf patch

Tradias committed:
protobuf: Adjust braft, libtorch, osgearth and pulsar-client-cpp to protobuf update

Tradias committed:
srpc: Adjust baseline version

Tradias committed:
libprotobuf-mutator: Fix msvc static runtime

Tradias committed:
protobuf: Adjust osgearth and paraview patches

Tradias committed:
srpc: Update to 0.10.1 and fix protobuf linkage

Tradias committed:
protobuf: Adjust removal of unused language backends and fix srpc get_filename_component error

Tradias committed:
protobuf: Fix depend cld3 library. Adjust srpc crt linkage. Remove duplicate versions of protobuf dependend ports

dg0yt commented on ports/braft/protobuf.patch:
```diff
@@ -0,0 +1,27 @@
+@@ -22,8 +22,8 @@ endif()
+ set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
+ 
+ include(FindThreads)
++find_package(protobuf CONFIG REQUIRED)
+ include(FindProtobuf)
```
Oops, Upstream should use `find_package(Foo)` instead of `include(FindFoo)`.

dg0yt commented on ports/braft/protobuf.patch:
```diff
@@ -0,0 +1,27 @@
+index a851c00..1b6f6a9 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -22,8 +22,8 @@ endif()
+ set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
+ 
+ include(FindThreads)
++find_package(protobuf CONFIG REQUIRED)
+ include(FindProtobuf)
+-
+ if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+     # require at least gcc 4.8
+     if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.8)
```
```suggestion
@@ -22,7 +22,7 @@ endif()
 set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
 
 include(FindThreads)
-include(FindProtobuf)
+find_package(Protobuf)
 
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     # require at least gcc 4.8
```
This will use the vcpkg cmake wrapper and the standard find module,

dg0yt commented on ports/braft/protobuf.patch:
```diff
@@ -0,0 +1,27 @@
+ if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+     # require at least gcc 4.8
+     if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.8)
+diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
+index ff435a2..59b10ae 100644
+--- a/src/CMakeLists.txt
++++ b/src/CMakeLists.txt
+@@ -10,7 +10,7 @@ include_directories(${CMAKE_CURRENT_BINARY_DIR})
+ include_directories(${CMAKE_SOURCE_DIR}/src)
+ 
+ add_library(OBJ_LIB OBJECT ${SOURCES})
+-
++target_link_libraries(OBJ_LIB PUBLIC protobuf::libprotobuf)
+ set_property(TARGET ${OBJ_LIB} PROPERTY POSITION_INDEPENDENT_CODE 1)
+ if (BUILD_SHARED_LIBS)
+ add_library(braft-shared SHARED $<TARGET_OBJECTS:OBJ_LIB>)
```
```suggestion
```
Obsolete with above change.

Tradias replied:
It is not. Propagation of `cxx_std_14/17` and abseil library linkage is still required:

```
/home/dh/vcpkg-test/build-clang/vcpkg_installed/x64-linux-release/include/absl/base/policy_checks.h:79:2: error: #error "C++ versions less than C++14 are not supported."
   79 | #error "C++ versions less than C++14 are not supported."
      |  ^~~~~
```

dg0yt reviewed (commented):
I see many changes to `find_package` config mode and imported targets.
But I don' see many `find_dependency` additions to exported cmake config.
Is this incomplete? Is it necessary to switch at all?

Tradias committed:
protobuf: Fix dependency propagation through cmake config file in braft and libprotobuf-mutator

Tradias committed:
protobuf: Fix cld3, brpc, ecal and srpc osx builds

Tradias committed:
protobuf: Fix ecal and srpc patches

Tradias committed:
mysql-connector-cpp: Add protobuf patch

Tradias committed:
marble: Patch accidental dependency on protobuf

Tradias commented:
I do not know why the GMP build is failing on Windows, I did not touch it or any of it's dependencies (it has none). From what I know GMP is not really supposed to be used on Windows anyways, MPIR should be used instead.

The `failure logs for x64-uwp\gmp\build-x64-windows-dbg-err.log` error message:
```
./libtool: eval: line 10827: syntax error near unexpected token `|'
./libtool: eval: line 10827: `dumpbin.exe -symbols -headers .libs/libgmp.la-2.obj  |  | /usr/bin/sed -e '/^[BCDGRS][ ]/s/.*[ ]\([^ ]*\)/\1,DATA/' | /usr/bin/sed -e '/^[AITW][ ]/s/.*[ ]//' | sort | uniq > .libs/gmp.exp'
make[2]: *** [Makefile:886: libgmp.la] Error 2
make[1]: *** [Makefile:1001: all-recursive] Error 1
make: *** [Makefile:791: all] Error 2
```

dg0yt commented:
> I do not know why the GMP build is failing on Windows, I did not touch it or any of it's dependencies (it has none). From what I know GMP is not really supposed to be used on Windows anyways, MPIR should be used instead.

No, MPIR is kept building, but not updated. The gmp build suffers from an unexpected behavioral change of the msys2 environment, however there seems to no x64-windows user available to nail it down. (https://github.com/microsoft/vcpkg/issues/36639#issuecomment-1938105619, https://github.com/microsoft/vcpkg/pull/36706)

luzpaz commented:
Note: current upstream stable is 1.62.0   
https://github.com/grpc/grpc/releases/tag/v1.62.0

Tradias commented:
@luzpaz I know, one step at a time please. This pr is already massive.

dg0yt commented:
The gmp issue is fixed, you must push a new state to trigger a  new CI run.

BillyONeal commented:
/azp run

BillyONeal commented:
Resetting Pipelines because I see binary caching isn't working. I just tried rolling SAS tokens.

BillyONeal commented:
/azp run

BillyONeal commented:
sorry :(

BillyONeal commented:
/azp run

BillyONeal commented:
Something is wrong with the SAS token making script :(

BillyONeal commented:
OK, everything is working now, sorry for the spam!

BillyONeal commented:
https://github.com/microsoft/vcpkg/pull/36938 is the result of the spam. Sorry again! 😅

MonicaLiu0311 commented:
Please get [failure logs here](https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_apis/build/builds/99805/artifacts?artifactName=failure%20logs%20for%20x64-windows-static&api-version=7.1&%24format=zip)
```
Installing 600/890 hpx[bzip2,core,mpi,snappy,zlib]:x64-windows-static@1.9.1#1...
Building hpx[bzip2,core,mpi,snappy,zlib]:x64-windows-static@1.9.1#1...
-- Note: hpx only supports dynamic library linkage. Building dynamic library.
CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:25 (message):
  Refusing to build unexpected dynamic library against the static CRT.

      If this is desired, please configure your triplet to directly request this configuration.
Call Stack (most recent call first):
  ports/hpx/portfile.cmake:2 (vcpkg_check_linkage)
  scripts/ports.cmake:170 (include)


error: building hpx:x64-windows-static failed with: BUILD_FAILED
Elapsed time to handle hpx:x64-windows-static: 45 ms
```

BillyONeal commented on ports/brpc/fix-build.patch:
```diff
@@ -132,14 +140,15 @@ index fbcc7cc5..95037a0d 100644
-+        target_link_libraries(brpc-shared ${THRIFT_LIB})
+     if(WITH_THRIFT)
+-        target_link_libraries(brpc-shared ${THRIFT_LIB})
+-        check_thrift_version(brpc-shared)
```
Is it possible / does it make sense to leave this version check in? As long as it's a strict > check it can be OK...

Tradias replied:
The check does not work:

```
brpc compile with thrift protocol
CMake Error at src/CMakeLists.txt:44 (execute_process):
  execute_process error getting child return code: No such file or directory
Call Stack (most recent call first):
  src/CMakeLists.txt:69 (check_thrift_version)
```
They try to determine thrift version by executing `thrift --version`.

BillyONeal commented on ports/brpc/fix-warnings.patch:
```diff
@@ -0,0 +1,46 @@
+     if (k != sizeof(version_and_len_buf) ) {
+         return butil::Status(-1, "Fail to copy %" PRIu64 " bytes from body",
+-                             sizeof(version_and_len_buf));
++                             (uint64_t)sizeof(version_and_len_buf));
```
This looks upstreamable, has upstream been notified?

BillyONeal commented on ports/brpc/protobuf.patch:
```diff
@@ -0,0 +1,235 @@
+     return NULL;
+ }
+ 
+-static void BaiduStreamingLogHandler(google::protobuf::LogLevel level,
+-                                     const char* filename, int line,
+-                                     const std::string& message) {
+-    switch (level) {
+-    case google::protobuf::LOGLEVEL_INFO:
+-        LOG(INFO) << filename << ':' << line << ' ' << message;
+-        return;
+-    case google::protobuf::LOGLEVEL_WARNING:
+-        LOG(WARNING) << filename << ':' << line << ' ' << message;
+-        return;
+-    case google::protobuf::LOGLEVEL_ERROR:
+-        LOG(ERROR) << filename << ':' << line << ' ' << message;
+-        return;
+-    case google::protobuf::LOGLEVEL_FATAL:
+-        LOG(FATAL) << filename << ':' << line << ' ' << message;
+-        return;
+-    }
+-    CHECK(false) << filename << ':' << line << ' ' << message;
+-}
+ 
+ static void GlobalInitializeOrDieImpl() {
+     //////////////////////////////////////////////////////////////////
+@@ -327,7 +308,6 @@ static void GlobalInitializeOrDieImpl() {
+     }
+ 
+     // Make GOOGLE_LOG print to comlog device
+-    SetLogHandler(&BaiduStreamingLogHandler);
```
I think we need upstream to OK this. There has to be some reason they were overriding this....

Tradias replied:
They are not the only ones affected by the removal of protobuf logging API. The other library also just removed their override. But I agree that upstream can be informed, I wonder what we do in the meantime?

Tradias replied:
It has been upstreamed

BillyONeal commented on ports/brpc/protobuf.patch:
```diff
@@ -0,0 +1,235 @@
+ 
+ // Since kDefaultTotalBytesLimit is private, we need some hacks to get the limit.
+ // Works for pb 2.4, 2.6, 3.0
+-#define private public
+ #include <google/protobuf/io/coded_stream.h>
+-const int PB_TOTAL_BYETS_LIMITS_RAW =
+-    google::protobuf::io::CodedInputStream::kDefaultTotalBytesLimit;
++const int PB_TOTAL_BYETS_LIMITS_RAW = INT_MAX;
```
To other reviewers:

Note 'BYETS' is misspelled in the original code.

Note also https://github.com/protocolbuffers/protobuf/issues/1927#issuecomment-284582993

TL;DR: This is OK

Tradias replied:
This has been upstreamed

BillyONeal commented on ports/cld3/unofficial-cld3Config.cmake:
```diff
@@ -1,4 +1,4 @@
 include(CMakeFindDependencyMacro)
-find_dependency(Protobuf)
+find_dependency(protobuf)
```
```suggestion
find_dependency(protobuf CONFIG)
```

because `find_package(protobuf CONFIG REQUIRED)`

dg0yt replied:
And literally `Protobuf`also for config mode, cf. https://github.com/microsoft/vcpkg/discussions/36934#discussioncomment-8598486
(vcpkg tool usage heuristics...)

BillyONeal commented on scripts/ci.baseline.txt:
```diff
@@ -761,9 +761,18 @@ mpir:x64-osx=skip
-mysql-connector-cpp:arm-neon-android=fail
-mysql-connector-cpp:arm64-android=fail
-mysql-connector-cpp:x64-android=fail
+mysql-connector-cpp:arm-neon-android=skip
```
Can you comment why these are all skip now? That's disconcerting given the number of changes made to that port in this PR.

Ditto shogun below (although that one was at least already skip for many platforms)

Tradias replied:
Yes, I wanted to update my pr discription with all the changes and why ... just so many now ^^.
Anyways, shogun does not work with v5 of bitsery but I patched it enough so that users who override bitsery to v4.x can use it. Bitsery dependency was introduced as I updated the version of shogun, but older versions had several very difficult to patch protobuf issues.
`mysql-connector-cpp` has the same issue as brpc: removal of `protobuf::SetLogHandler` and maybe more issues after that.

Tradias replied:
Reverted

BillyONeal reviewed (commented):
I can't approve due to the ci.baseline.txt skips but most the rest of this looks good... THANK YOU for digging into this mess!

Tradias committed:
protobuf: Change patches of several ports from find_package(protobuf back to find_package(Protobuf.
Re-enable mysql-connector-cpp ci build. Fix port-version of cld3

ilya-lavrenov commented on ports/openvino/007-find-protobuf.patch:
```diff
@@ -0,0 +1,12 @@
+         # see https://protobuf.dev/support/version-support/ and
+         # https://github.com/protocolbuffers/protobuf/commit/d61f75ff6db36b4f9c0765f131f8edc2f86310fa
+         find_package(Protobuf 4.22.0 QUIET CONFIG)
++        set(protobuf_config CONFIG)
```
why is it needed? we can see that on line 12 it's already set to `CONFIG` for vcpkg

Tradias replied:
The line above actually finds protobuf butl ater in `OpenVINOConfig.cmake.in` we need `protobuf_config` to be set however, otherwise this block:

```cmake
macro(_ov_find_protobuf_frontend_dependency)
    set(_OV_ENABLE_SYSTEM_PROTOBUF "@ENABLE_SYSTEM_PROTOBUF@")
    set(_OV_PROTOBUF_PACKAGE_CONFIG "@protobuf_config@")
    if(_OV_ENABLE_SYSTEM_PROTOBUF)
        _ov_find_dependency(Protobuf @Protobuf_VERSION@ EXACT ${_OV_PROTOBUF_PACKAGE_CONFIG})
    endif()
```
will call `find_dependency` without `CONFIG`.

ilya-lavrenov replied:
ported to upstream https://github.com/openvinotoolkit/openvino/pull/23460

Tradias committed:
protobuf: Patch mysql-connector-cpp and clean up openvino patch

dg0yt commented:
IMO there should be an attempt to modify the `protobuf` cmake wrapper as much as much as necessary to remove the need for most
~~~diff
-find_dependency(Protobuf)
+find_package(Protobuf CONFIG)
~~~
which triggers more patching (`find_dependency` etc.). Isn't this the whole point of having the wrappers?

Tradias committed:
glog: FIx braft and brpc after update

Tradias committed:
protobuf: Attempt to fix mysql-connector-cpp protobuf executable not found. Restore openvino patch

Tradias committed:
protobuf: Fix mysql-connector-cpp when using protobuf with zlib feature

Tradias committed:
protobuf: Fix braft patch. Re-add port-version to marble. Fix ZLIB in mysql-connector-cpp

Tradias committed:
protobuf: Adjust mysql-connector-cpp patch to properly work with zlib feature. Fix brpc find openssl on darwin systems

Tradias committed:
brpc: Patch out #warning because OSX compiles with -Werror

kkuehl commented:
I've been watching this PR for a long time, obviously a massive amount of work has gone into this, thank you!

dg0yt commented:
Please make the PR title carry also the protobuf update.

coryan commented on ports/protobuf/vcpkg.json:
```diff
@@ -1,15 +1,16 @@
   "name": "protobuf",
-  "version": "3.21.12",
-  "port-version": 1,
+  "version": "3.25.1",
```
This should be 4.25.1.  Protobuf uses an unusual versioning scheme, with different major versions for different languages.  In this case, v25.x is 4.25.x for C++ and 3.25.x for other languages:

https://protobuf.dev/support/version-support/

coryan commented on ports/utf8-range/vcpkg.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "utf8-range",
-  "version-date": "2023-11-09",
+  "version": "3.25.1",
```
This should also be 4.25.1: https://protobuf.dev/support/version-support/

coryan reviewed (commented):
A few more nits on Protobuf versions.

Tradias commented on ports/grpc/00001-fix-uwp.patch:
```diff
@@ -50,18 +50,18 @@ index 25990a5d8a..4bec4e1e72 100644
                                           const char* reg_key_path,
                                           const char* reg_key_name) {
-+#if !defined(WINAPI_FAMILY) || WINAPI_FAMILY == WINAPI_FAMILY_DESKTOP_APP
++  #if !defined(WINAPI_FAMILY) || WINAPI_FAMILY == WINAPI_FAMILY_DESKTOP_APP
```
I didn't make this change it comes from this pr: https://github.com/microsoft/vcpkg/pull/31159
The space doesn't matter though, so I won't change.

dg0yt commented:
Port cld3 is already fixed now.

Tradias committed:
cld3: Bump port version

Tradias committed:
grpc: Add patch for abseil 20240116.1

Tradias committed:
abseil: Fix version in abseil_dll.pc

Tradias committed:
shogun: Remove unneeded patches

okravets commented:
Could this PR to get merged?
Thank you !

dg0yt commented on ports/abseil/abseil-dll-pc.patch:
```diff
@@ -0,0 +1,13 @@
+ Description: Abseil DLL library\n\
+ URL: https://abseil.io/\n\
+-Version: ${absl_VERSION}\n\
++Version: head\n\
```
I don't think this is a good idea: It will breal version expression (`abseil >= x.y.z`).

Tradias replied:
Reverted

dg0yt commented on ports/abseil/portfile.cmake:
```diff
@@ -5,9 +5,11 @@ endif()
     REPO abseil/abseil-cpp
-    REF "${VERSION}"
-    SHA512 41504899ac4fd4a6eaa0a5fdf27a7765ec81962fb99b6a07982ceed32c5289e9eb12206c83a70fd44c5c3e1b96c2bfa160eb12f1dbbb45f1109d632c7690de90
+    REF d5e426097680beafbf72d6c88f50a680736ec770 # 03/04/2024
```
... so this isn't really 20240116.1.

Tradias replied:
I reverted to upgrade and added a gigantic official patch to address the openvino MSVC compile error.

Tradias committed:
abseil: Revert abseil back to 20240116.1 and add patch to fix openvino compile error

MonicaLiu0311 commented:
grpc
-
All features are tested successfully in the following triplet:
```
x86-windows
x64-windows
x64-windows-static
```

The usage test passed on `x64-windows` (header files found):
```
grpc provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(gRPC CONFIG REQUIRED)
  # note: 8 additional targets are not displayed.
  target_link_libraries(main PRIVATE gRPC::gpr gRPC::grpc gRPC::grpc++ gRPC::grpc++_alts)
```

protobuf
-
All features are tested successfully in the following triplet:
```
x86-windows
x64-windows
x64-windows-static
```

JavierMatosD reviewed (commented):
Most of the changes in this PR are good to go but I need explanations for the various patches. Why are they needed? In general, if the patches aren't required to build the port then they shouldn't be added here. Bugs should be reported to upstream. I need explanations for the following patches:

- `ports/brpc/fix-warnings.patch`
- `ports/brpc/protobuf.patch`
- `ports/ignition-transport4/fix-protobuf.patch`
- `ports/ignition-transport8/fix-protobuf.patch`
- `ports/ignition-transport9/fix-protobuf.patch`
- `ports/libprotobuf-mutator/dont-use-syntax.patch`
- `ports/libprotobuf-mutator/switch-to-requires_utf8_validation.patch`
- `ports/mysql-connector-cpp/protobuf-source.patch`
- `ports/protobuf/fix-default-proto-file-path.patch`
- `ports/shogun/eigen-3.4.patch`
- `ports/shogun/fmt.patch`

I see that you've consolidated some patches as well. If any of these patches were present before your changes then please lmk.

I also see that `shogun` is marked as fail in `ci.baseline.txt` in all platforms? Every port should be tested in at least one platform barring extremely niche circumstances. 

Finally, the versions selected don't seem to correlate to anything obvious. For example, `upb` changes the version scheme from version-date to version: 3.25.1. Upstream reports v26.1 (assuming you want to update after this PR) but https://github.com/microsoft/vcpkg/pull/35781/files#r1528420738 suggests it should be 4.25.1? 

Looking at https://protobuf.dev/support/version-support/

<img width="651" alt="image" src="https://github.com/microsoft/vcpkg/assets/44559473/d4c7a035-efdd-4543-916b-60ad713431f2">

Indicates maybe it should be 4.25.1

Tradias commented on ports/mysql-connector-cpp/protobuf-source.patch:
```diff
@@ -0,0 +1,61 @@
+diff --git a/cdk/protocol/mysqlx/protocol.cc b/cdk/protocol/mysqlx/protocol.cc
+index d4589c27..c61deca4 100644
```
This is due to a breaking change in protobuf. It is unclear whether upstream has any interest in supporting this protobuf version any time soon. Additionally, it is unclear whether this patch correctly migrates the breaking change.

Tradias replied:
For the last part I created https://github.com/protocolbuffers/protobuf/issues/16914

Tradias replied:
Seeing that brpc decided to roll with the simple removal of `protobuf::SetLogHandler` and Protobuf responded  that exceptions are not supported in the new logging system, I don't see any other way of migrating to protobuf 4 other than what I have done here.

Tradias commented on scripts/ci.baseline.txt:
```diff
@@ -1037,15 +1037,19 @@ septag-sx:x64-android=fail
-shogun:x64-uwp            = skip
-shogun:x64-windows        = skip
-shogun:x64-windows-static = skip
+shogun:arm-neon-android=skip
```
No, current shogun version is incompatible with new protobuf version and HEAD adds a dependency on bitsery 4 which is incompatible with vcpkg (bitsery 5).

Tradias replied:
I patched out bitsery and restored ci.baseline

Tradias commented on ports/upb/fix-cmake.patch:
```diff
@@ -0,0 +1,202 @@
+   reflection_internal)
+ 
+ 
++target_sources(upb PRIVATE
++  ../base/status.c
++  ../hash/common.c
++  ../io/tokenizer.c
++  ../io/chunked_input_stream.c
++  ../io/chunked_output_stream.c
++  ../json/decode.c
++  ../json/encode.c
++  ../lex/atoi.c
++  ../lex/round_trip.c
++  ../lex/strtod.c
++  ../lex/unicode.c
++  ../mem/alloc.c
++  ../mem/arena.c
++  ../message/accessors.c
++  ../message/copy.c
++  ../message/array.c
++  ../message/map.c
++  ../message/map_sorter.c
++  ../message/message.c
```
We too, unfortunately a simple 'exclude `*_test.c`' GLOB won't do.

Tradias commented on ports/shogun/portfile.cmake:
```diff
@@ -56,17 +53,17 @@ vcpkg_cmake_configure(
         -DCMAKE_DISABLE_FIND_PACKAGE_OpenMP=TRUE
         -DINSTALL_TARGETS=shogun-static
         ${extra_opts}
+        -DCMAKE_CXX_STANDARD=14 # protobuf
```
Protobuf requires C++14 and shogun does not use modern target-based approach which means `cxx_std_14` is not propagated from `protobuf::libprotobuf`. Since shogun is static library only this was the easiest way to get it to work, rather than patching it. Users of shogun must have been used to adding `protobuf::libprotobuf` to their target link libraries already anyways.

Tradias commented:
> ports/brpc/fix-warnings.patch

Since vcpkg's Mac pipeline compiles with `-Werror` for some reasons, these warnings had to be addressed. They can certainly be upstreamed.

> ports/brpc/protobuf.patch

Fixes breaking changes introduced by this version of protobuf. I do not know whether brpc has any interest in supporting this version of protobuf which means changes may or may not be upstreamable.

> ports/ignition-transport4/fix-protobuf.patch
> ports/ignition-transport8/fix-protobuf.patch
> ports/ignition-transport9/fix-protobuf.patch

Backport protobuf breaking changes fixes from more recent major release.

> ports/libprotobuf-mutator/dont-use-syntax.patch
> ports/libprotobuf-mutator/switch-to-requires_utf8_validation.patch

Unreleased official commits to fix compatibility with this version of protobuf. They have since been released (1.3), I removed them and updated the port.

> ports/mysql-connector-cpp/protobuf-source.patch

See comment on file.

> ports/protobuf/fix-default-proto-file-path.patch

This patch has been there before.

> ports/shogun/eigen-3.4.patch

This patch has been there before, but I consolidated it with the CMake fix.

> ports/shogun/fmt.patch

Fix breaking change caused by new version of fmt. Certainly upstreamable.

Tradias committed:
libprotobuf-mutator: Update to 1.3

Tradias committed:
protobuf: Adjust version to 4.25.1 and reduce upb patch

Tradias committed:
paraview: Bump port-version

Jeremy-Boyle commented:
Shouldn't the version be v5.x ?

No mention of 4.x public support .

Tradias commented:
@Jeremy-Boyle Not yet, this is v25.1 which corresponds to 4.25.1. See here https://protobuf.dev/support/version-support/

mathisloge commented:
FYI brpc has released a version which is compatible with protobuf >22.0 https://github.com/apache/brpc/releases/tag/1.9.0

Jeremy-Boyle commented:
> @Jeremy-Boyle Not yet, this is v25.1 which corresponds to 4.25.1. See here https://protobuf.dev/support/version-support/

Okay, thank you ! Im assuming a follow-up PR will have to be done to get inline with the latested supported version per the website. Thanks !

Tradias committed:
shogun: Remove bitsery and restore its ci.baseline

Tradias commented:
@mathisloge Good news, but v22 is still far away from v25. :)

@JavierMatosD Under what conditions will this pr be undrafted? I patched shogun further (removed bitsery) and restored its ci.baseline. It is now build successfully on some platforms.

dg0yt commented:
> Under what conditions will this pr be undrafted?

When you think it "ready for review", press the button.

RightFS commented:
I try this branch, there's a little problem.

I am in vcpkg manifest mode, and cross compiling arm64-Android on arm64-MacOS, protobuf 4.25.1 is right, but it will install 3.21.12#2 for `protoc`,  anything more I forgot ?

Now I can only use the classic mode, it works good.

```
The following packages will be rebuilt:
  * protobuf:arm64-osx@3.21.12#2 -- /Users/666/.vcpkg-clion/protobuf_vcpkg/buildtrees/versioning_/versions/protobuf/e189106689c1983cd5e19363fa48f35d3d3f419b
The following packages will be built and installed:
  * abseil:arm64-android@20240116.2 -- /Users/666/.vcpkg-clion/protobuf_vcpkg/buildtrees/versioning_/versions/abseil/9af16c8547b0a3d6c5886384b9b18934353c9d6d
    protobuf:arm64-android@4.25.1 -- /Users/666/.vcpkg-clion/protobuf_vcpkg/buildtrees/versioning_/versions/protobuf/c61b3b1e6a9761a6d6f52f90407c400a9cc954a8
  * utf8-range:arm64-android@2023-11-09 -- /Users/666/.vcpkg-clion/protobuf_vcpkg/buildtrees/versioning_/versions/utf8-range/dd54a8c4b774935b2c164d0902185fa56a8694bb
```

Tradias committed:
protobuf: Update versions file

Neumann-A commented on ports/ecal/portfile.cmake:
```diff
@@ -16,6 +16,7 @@ vcpkg_from_github(
         0005-remove-install-prefix-macro-value.patch
         0006-use-find_dependency-in-cmake-config.patch
         0007-allow-static-build-of-core.patch
+        0008-protobuf-linkage.patch
```
why is this patch necessary?

Tradias replied:
No longer necessary with vcpkg-cmake-wrapper in protobuf

Neumann-A commented on ports/arcus/vcpkg.json:
```diff
@@ -1,7 +1,6 @@
   "name": "arcus",
-  "version-semver": "4.10.0",
-  "port-version": 2,
+  "version-semver": "4.13.2",
```
This could be its own PR.

Neumann-A commented on ports/braft/protobuf.patch:
```diff
@@ -0,0 +1,23 @@
+ 
+ include(FindThreads)
+-include(FindProtobuf)
++find_package(Protobuf CONFIG REQUIRED)
```
I don't see why you are using CONFIG here.

Tradias replied:
CMake has a build-in https://cmake.org/cmake/help/latest/module/FindProtobuf.html which does not setup abseil linkage properly.

Neumann-A replied:
If that is the case you install a `vcpkg-cmake-wrapper.cmake` and fix that. You do not patch every downstream consumer.

dg0yt replied:
On March 5, I proposed https://github.com/microsoft/vcpkg/pull/35781#issuecomment-1979072441...

Tradias replied:
I didn't exactly know what the wrapper does, sorry for not asking. I added `CONFIG` to the protobuf wrapper now, it reduces some of the patching (cld3, gamenetworkingsockets, libosmscout, libprotobuf-mutator, osgearth, pulsar-client, ecal) but also adds some (arcus, cartographer).

dg0yt replied:
No problem.
There is one problem which the wrapper can't solve: When you switch a port from LIBRARIES variable to the imported target, you must add `find_dependency` to exported config.
This could be avoided by keeping variables free of targets, but often this is (too) hard.

Tradias replied:
> you must add find_dependency to exported config.

I don't understand why this is a `must`, it is merely an inconvenience. Users can still consume these ports by adding the missing `find_package` calls for the dependencies themselves.

> This could be avoided by keeping variables free of targets, but often this is (too) hard.

The way I changed protobuf's cmake-wrapper I no longer set these variables at all, not sure how it would help when most libs do `target_link_libraries(target ${PROTOBUF_LIBRARIES})` anyways.

dg0yt replied:
> I don't understand why this is a `must`, it is merely an inconvenience. Users can still consume these ports by adding the missing `find_package` calls for the dependencies themselves.

It is not a `must` to have cmake packages at all... User can just use `find_library` directly.

Seriously, it is a `must` at least in vcpkg that `find_package` cares for transitive dependencies.
- It is a fair requirement: Each (one!) package knows its direct requirements at build time. 
- It is not only inconvenient but also inefficient if (many!) users would need to work out networks of transitive dependencies.

Neumann-A commented on ports/braft/protobuf.patch:
```diff
@@ -0,0 +1,23 @@
+     message(FATAL_ERROR "Fail to find brpc")
+ endif()
+ 
+-if (NOT PROTOBUF_PROTOC_EXECUTABLE)
++set(PROTOBUF_LIBRARY protobuf::libprotobuf)
++if (0)
```
Seems unnecessary. PROTOBUF_PROTOC_EXECUTABLE should get set by FindProtobuf alternatively it could be set from the `portfile.cmake`

Tradias replied:
The important part here is `set(PROTOBUF_LIBRARY protobuf::libprotobuf)`. `PROTOBUF_PROTOC_EXECUTABLE` is untouched due to the `if(false) `.

Neumann-A commented on ports/gamenetworkingsockets/portfile.cmake:
```diff
@@ -20,6 +20,7 @@ vcpkg_cmake_configure(
         -DBUILD_TOOLS=OFF
         -DUSE_CRYPTO=${CRYPTO_BACKEND}
         -DUSE_CRYPTO25519=${CRYPTO_BACKEND}
+        -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=on
```
Again why?

Tradias replied:
```
   Creating library src\GameNetworkingSockets.lib and object src\GameNetworkingSockets.exp
steamnetworkingsockets_certstore.cpp.obj : error LNK2001: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > * __cdecl absl::lts_20240116::log_internal::MakeCheckOpString<__int64,__int64>(__int64,__int64,char const *)" (??$MakeCheckOpString@_J_J@log_internal@lts_20240116@absl@@YAPEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_J0PEBD@Z)
steamnetworkingsockets_certstore.cpp.obj : error LNK2001: unresolved external symbol "public: __cdecl absl::lts_20240116::log_internal::LogMessageFatal::~LogMessageFatal(void)" (??1LogMessageFatal@log_internal@lts_20240116@absl@@QEAA@XZ)
steamnetworkingsockets_certstore.cpp.obj : error LNK2001: unresolved external symbol "public: __cdecl absl::lts_20240116::log_internal::LogMessageFatal::LogMessageFatal(char const *,int,class absl::lts_20240116::string_view)" (??0LogMessageFatal@log_internal@lts_20240116@absl@@QEAA@PEBDHVstring_view@23@@Z)
steamnetworkingsockets_stats.cpp.obj : error LNK2001: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > * __cdecl absl::lts_20240116::log_internal::MakeCheckOpString<unsigned __int64,unsigned __int64>(unsigned __int64,unsigned __int64,char const *)" (??$MakeCheckOpString@_K_K@log_internal@lts_20240116@absl@@YAPEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_K0PEBD@Z)
steamnetworkingsockets_stats.cpp.obj : error LNK2001: unresolved external symbol "private: void __cdecl absl::lts_20240116::log_internal::LogMessage::CopyToEncodedBuffer<0>(class absl::lts_20240116::string_view)" (??$CopyToEncodedBuffer@$0A@@LogMessage@log_internal@lts_20240116@absl@@AEAAXVstring_view@23@@Z)
steamnetworkingsockets_messages_certs.pb.cc.obj : error LNK2001: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > * __cdecl absl::lts_20240116::log_internal::MakeCheckOpString<void const *,void const *>(void const *,void const *,char const *)" (??$MakeCheckOpString@PEBXPEBX@log_internal@lts_20240116@absl@@YAPEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PEBX0PEBD@Z)
steamnetworkingsockets_messages.pb.cc.obj : error LNK2001: unresolved external symbol "public: __cdecl absl::lts_20240116::log_internal::LogMessageFatal::LogMessageFatal(char const *,int)" (??0LogMessageFatal@log_internal@lts_20240116@absl@@QEAA@PEBDH@Z)

bin\GameNetworkingSockets.dll : fatal error LNK1120: 7 unresolved externals
```

Tradias committed:
protobuf: Include CONFIG in vcpkg-cmake-wrapper

Tradias committed:
protobuf: Fix find version in arcus and cartographer

Tradias committed:
cartographer: Fix absolute path issue and disable compilation of test lib

Tradias committed:
brpc/pulsar-cpp-client: Patch warnings caused by abseil dynamic_annotations.h and restore disable warnings patch

BillyONeal commented:
```
Exception ignored in: <function BaseFileLock.__del__ at 0x00000259484FEB60>
Traceback (most recent call last):
  File "D:\downloads\tools\python\python-3.11.8-x64\Lib\site-packages\filelock\_api.py", line 365, in __del__
    self.release(force=True)
  File "D:\downloads\tools\python\python-3.11.8-x64\Lib\site-packages\virtualenv\util\lock.py", line 34, in release
    with self.thread_safe:
         ^^^^^^^^^^^^^^^^
AttributeError: '_CountedFileLock' object has no attribute 'thread_safe'
Exception ignored in: <function BaseFileLock.__del__ at 0x00000259484FEB60>
Traceback (most recent call last):
  File "D:\downloads\tools\python\python-3.11.8-x64\Lib\site-packages\filelock\_api.py", line 365, in __del__
    self.release(force=True)
  File "D:\downloads\tools\python\python-3.11.8-x64\Lib\site-packages\virtualenv\util\lock.py", line 34, in release
    with self.thread_safe:
         ^^^^^^^^^^^^^^^^
AttributeError: '_CountedFileLock' object has no attribute 'thread_safe'
Exception ignored in: <function BaseFileLock.__del__ at 0x00000259484FEB60>
Traceback (most recent call last):
  File "D:\downloads\tools\python\python-3.11.8-x64\Lib\site-packages\filelock\_api.py", line 365, in __del__
    self.release(force=True)
  File "D:\downloads\tools\python\python-3.11.8-x64\Lib\site-packages\virtualenv\util\lock.py", line 34, in release
    with self.thread_safe:
         ^^^^^^^^^^^^^^^^
AttributeError: '_CountedFileLock' object has no attribute 'thread_safe'
```

Would updating the python version selected fix this?

Tradias commented:
@BillyONeal I didn't touch that port or any of its dependencies from what I can tell. Please fix on master then I will rebase.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal commented:
> @BillyONeal I didn't touch that port or any of its dependencies from what I can tell. Please fix on master then I will rebase.

As far as I can tell master isn't experiencing this failure right now so I'm not sure what's going on. Just going to kick it again and see if it works out

Tradias committed:
Merge branch 'master' into grpc-1-60-0

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal committed:
[protobuf] Skip unchanged vcpkg_replace_string.

BillyONeal commented:
Thanks for the extreme update!

Tradias commented:
@BillyONeal I am glad we finally got it merged, took only 7 month 🙈

dg0yt commented:
But #39459...

cenit commented:
this pr silently broke protobuf in opencv, the fix is include in my opencv 4.9 pr... (which also includes a very nice feature to avoid other silent breakings like this)

davidhunter22 commented:
Sadly the /permissive work around does not always work. On the upside this seem to have been fixed or at least worked  around in protobuf 26.0. See https://github.com/microsoft/vcpkg/issues/39459 for details.
So protobuf 25.0 and MSVC seems to be pretty toxic!

yonik commented:
Thanks for this massive amount of work @Tradias! 
Unfortunately, I've been hitting ASAN errors since the update, and reproduced it with the smallest example I could come up with:
https://github.com/yonik/grpc_example

Tradias commented:
@yonik I didn't look into your example more closely but immediately noticed the lack of https://github.com/grpc/grpc/pull/22325. Since you are using the default vcpkg triplet, gRPC will be compiled without sanitizer enabled. You must either create a with sanitizer enabled or use `GRPC_ASAN_SUPPRESSED`.

yonik commented:
Thanks @Tradias, building my example with GRPC_ASAN_SUPPRESSED defined didn't seem to work, but rebuilding everything in vcpkg with ASAN did!
